File: demo/src/scala/tuiexamples/PopupExample.scala
package tuiexamples

import tui._
import tui.crossterm.CrosstermJni
import tui.widgets._

object PopupExample {
  case class App(
      var show_popup: Boolean = false
  )

  def main(args: Array[String]): Unit =
    withTerminal { (jni, terminal) =>
      val app = App()
      run_app(terminal, app, jni);
    }

  def run_app(terminal: Terminal, app: App, jni: CrosstermJni): Unit =
    while (true) {
      terminal.draw(f => ui(f, app))

      jni.read() match {
        case key: tui.crossterm.Event.Key =>
          key.keyEvent.code match {
            case char: tui.crossterm.KeyCode.Char if char.c() == 'q' => return
            case char: tui.crossterm.KeyCode.Char if char.c() == 'p' => app.show_popup = !app.show_popup
            case _                                                   => ()
          }
        case _ => ()
      }
    }

  def ui(f: Frame, app: App): Unit = {
    val text = if (app.show_popup) "Press p to close the popup" else "Press p to show the popup"
    Layout
      .detailed()(
        Constraint.Percentage(20) -> ParagraphWidget(
          text = Text.from(Span.styled(text, Style(addModifier = Modifier.SLOW_BLINK))),
          alignment = Alignment.Center,
          wrap = Some(ParagraphWidget.Wrap(trim = true))
        ),
        Constraint.Percentage(80) -> BlockWidget.noChildren(
          title = Some(Spans.nostyle("Content")),
          borders = Borders.ALL,
          style = Style.DEFAULT.bg(Color.Blue)
        )
      )
      .render(f.size, f.buffer)

    if (app.show_popup) {
      val block = BlockWidget.noChildren(title = Some(Spans.nostyle("Popup")), borders = Borders.ALL)
      val area = centered_rect(60, 20, f.size)
      f.renderWidget(ClearWidget, area); // this clears out the background
      f.renderWidget(block, area)
    }
  }

/// helper function to create a centered rect using up certain percentage of the available rect `r`
  def centered_rect(percent_x: Int, percent_y: Int, r: Rect): Rect = {
    val popup_layout = Layout.cached(
      area = r,
      constraints = Array(
        Constraint.Percentage((100 - percent_y) / 2),
        Constraint.Percentage(percent_y),
        Constraint.Percentage((100 - percent_y) / 2)
      ),
      margin = Margin.None,
      direction = Direction.Vertical,
      expandToFill = false
    )

    Layout.cached(
      area = popup_layout(1),
      constraints = Array(
        Constraint.Percentage((100 - percent_x) / 2),
        Constraint.Percentage(percent_x),
        Constraint.Percentage((100 - percent_x) / 2)
      ),
      margin = Margin.None,
      direction = Direction.Horizontal,
      expandToFill = false
    )(1)
  }
}

File: demo/src/java/tuiexamples/PopupExample.java
package tuiexamples;

import tui.Alignment;
import tui.Borders;
import tui.Color;
import tui.Constraint;
import tui.Direction;
import tui.Frame;
import tui.Layout;
import tui.Margin;
import tui.Modifier;
import tui.Rect;
import tui.Span;
import tui.Spans;
import tui.Style;
import tui.Terminal;
import tui.Text;
import tui.WithTerminal;
import tui.crossterm.CrosstermJni;
import tui.crossterm.Event;
import tui.crossterm.KeyCode;
import tui.widgets.BlockWidget;
import tui.widgets.ClearWidget;
import tui.widgets.ParagraphWidget;

public final class PopupExample {
  private PopupExample() {}

  public static final class App {
    public boolean showPopup;

    public App(boolean showPopup) {
      this.showPopup = showPopup;
    }
  }

  public static void main(String[] args) {
    WithTerminal.apply((jni, terminal) -> {
      App app = new App(false);
      runApp(terminal, app, jni);
      return null;
    });
  }

  public static void runApp(Terminal terminal, App app, CrosstermJni jni) {
    while (true) {
      terminal.draw(f -> ui(f, app));

      Event ev = jni.read();
      if (ev instanceof Event.Key key) {
        if (key.keyEvent().code() instanceof KeyCode.Char c) {
          if (c.c() == 'q') {
            return;
          } else if (c.c() == 'p') {
            app.showPopup = !app.showPopup;
          }
        }
      }
    }
  }

  public static void ui(Frame f, App app) {
    Rect size = f.size;

    Layout layout =
        new Layout(
            Direction.Vertical,
            new Margin(0, 0),
            new Constraint[] {new Constraint.Percentage(20), new Constraint.Percentage(80)}, true);
    Rect[] chunks = layout.split(size);

    String text = app.showPopup ? "Press p to close the popup" : "Press p to show the popup";
    ParagraphWidget paragraph =
        ParagraphWidget.empty(
                Text.from(
                    Span.styled(text, Style.empty().withAddModifier(Modifier.SLOW_BLINK))))
            .withAlignment(Alignment.Center)
            .withWrap(new ParagraphWidget.Wrap(true));
    f.renderWidget(paragraph, chunks[0]);

    BlockWidget block =
        BlockWidget.empty()
            .withTitle(Spans.nostyle("Content"))
            .withBorders(Borders.ALL)
            .withStyle(Style.DEFAULT.withBg(Color.Blue));
    f.renderWidget(block, chunks[1]);

    if (app.showPopup) {
      BlockWidget popup =
          BlockWidget.empty()
              .withTitle(Spans.nostyle("Popup"))
              .withBorders(Borders.ALL);
      Rect area = centeredRect(60, 20, size);
      f.renderWidget(ClearWidget.INSTANCE, area);
      f.renderWidget(popup, area);
    }
  }

  public static Rect centeredRect(int percentX, int percentY, Rect r) {
    Layout popupLayout =
        new Layout(
            Direction.Vertical,
            new Margin(0, 0),
            new Constraint[] {
              new Constraint.Percentage((100 - percentY) / 2),
              new Constraint.Percentage(percentY),
              new Constraint.Percentage((100 - percentY) / 2)
            },
            false);
    Rect[] popupChunks = popupLayout.split(r);

    Layout horizontal =
        new Layout(
            Direction.Horizontal,
            new Margin(0, 0),
            new Constraint[] {
              new Constraint.Percentage((100 - percentX) / 2),
              new Constraint.Percentage(percentX),
              new Constraint.Percentage((100 - percentX) / 2)
            },
            false);
    return horizontal.split(popupChunks[1])[1];
  }
}
